### ISSUE TITLE
get/set_rotation and get/set_yaw problems

### ISSUE THREAD

**Issue by pgimeno** [minetest/minetest#7927]
##### Issue type

 - Bug report
 
##### Minetest version

```
Minetest 5.0.0-dev-eb211788-dirty (Linux)
```

##### Summary

I finally got a chance to test the `set_rotation()` method introduced in #7395. Thanks Paramat for the Boats patch.

The problem is that the X angle is roll, Y is pitch and Z is yaw, but due to Minetest's weird axis conventions, yaw is rotating on the wrong axis, as it should rotate around the Y axis, which is the vertical one in Minetest's case, but it's currently rotating around the Z axis, which is the usual convention but is horizontal in Minetest. Pitch has a similar problem. Roll is working just fine, over the local X axis of the object. I don't know if that matches MT's convention for the forward vector.

It should be possible to convert the input angles of `set_rotation()` into a matrix or quaternion, and then convert that matrix or quaternion back to Eulers but in the order expected by Irrlicht. Similarly for `get_rotation()`. There will be some loss of precision in the process, but hopefully small enough to be manageable.

That, however, will open another can of worms, namely that these kinds of calculations often generate rotations that are adjacent, but whose values are wildly different. For example, in degrees, `0,a,0` is equivalent to `180,b,180` for `a=90-b`, and it's often the case that the calculations that handle the conversion will change from 0 to 180 when reaching a 90° pitch (that's the gimbal lock angle, incidentally). However, the interpolator knows nothing about that equivalence, and will interpolate from 0 all the way to 180 in both axes, causing rotation jerks.

cc @CoderForTheBetter

**Comment by paramat** [comment]
? According to my boats patch Y is yaw, X is pitch and Z is an unintuitive roll.
If you're nor careful when testin gthe unintuitive roll can really mess up your perception of rotation. Start with using yaw and pitch only and you will see it works fine.

**Comment by pgimeno** [comment]
@paramat I'm completely sure about what I said. Try to imagine the boat as an airplane with the nose pointing to the starboard of the boat, the wings crossing the table perpendicularly, and the bottom of the plane where the boat engine is. Then try to imagine that the water is a vertical wall, and it will all make sense.

See this boat for reference. It is at 0,0,0 rotation.
![minetest-boat-as-plane](https://user-images.githubusercontent.com/4267396/49338333-fcc52e00-f61f-11e8-9ec2-55d17d0e5c9a.png)

With up/down (shift/space) you control the yaw over the Z axis, with left/right (A/D) the pitch, and with fwd/back (W/S) the roll.

Now, pitch and roll (the Y and X angles) give the illusion to be working as yaw and pitch over a different axis, but only as long as Z (the actual yaw) is zero. That's because their relative order is correct. Similarly, yaw and roll can give the illusion of working as pitch and roll too, for the same reason. But the actual yaw is the Z angle, the actual pitch is the Y angle, and the actual roll is the X angle.

**Comment by rubenwardy** [comment]
The correct behaviour is:

* X is pitch
* Y is yaw
* Z is roll

Think of it as a pin through that axis which the object rotates around

**Comment by pgimeno** [comment]
I agree that that's the desired behaviour. That's why the current behaviour is broken, which is the reason for me to filing this issue, before code that depends on current behaviour is written.

In intrinsic angles ("pins through the object", i.e. moving rotation axes) you first rotate over the yaw angle to select the desired heading, then over the pitch angle to select the desired bearing, and finally over the roll angle. In extrinsic angles (fixed rotation axes, which you can think of as a machine fixed on the ground that grabs the model, rotates it and releases it) you do it in the opposite order and obtain the same result.

There are 24 variants of Euler angles. Half of them are proper Euler angles, which are ABA rotations instead of ABC (where A, B and C stand for any of the X, Y and Z axes, and different letters indicate different axes), which we are not interested in because these don't actually represent roll/pitch/yaw. The other 12 are Tait-Bryan angles, which are of the form ABC and represent roll/pitch/yaw. Of these 12, half are intrinsic and half are extrinsic. The remaining 6 of each are the possible orderings of X, Y and Z.

No matter the ordering or the convention, in Tait-Bryan angles there will always be one axis behaving as yaw (purely extrinsic rotation over one of the world's axes), one axis behaving as roll (pure intrinsic rotation over one of the object's local axes) and one axis behaving as pitch (it's sandwiched between the others, so it's neither intrinsic nor extrinsic unless one of the others is zero).

Irrlicht chooses the intrinsic ZYX order (or equivalently, the extrinsic XYZ order). That's clear from the [sources](http://irrlicht.sourceforge.net/docu/matrix4_8h_source.html#l00825) (except for the sign convention, compare to the [Wikipedia table](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix)). But that doesn't fare well with Minetest's choice of world axes, where Z and Y are swapped, because Z is not vertical, therefore the yaw is not vertical. Hence this issue.

**Comment by paramat** [comment]
> Now, pitch and roll (the Y and X angles) give the illusion to be working as yaw and pitch over a different axis, but only as long as Z (the actual yaw) is zero. That's because their relative order is correct. Similarly, yaw and roll can give the illusion of working as pitch and roll too, for the same reason. But the actual yaw is the Z angle, the actual pitch is the Y angle, and the actual roll is the X angle.

Ok i see now. However the rotation PR simply applies Irrlicht 3D rotations as they work in Irrlicht and in MT, however unintuitive or yaw-pitch-roll-order-incorrect that may be due to our axis system, because attachment 3D rotations work this way too, so we did this for consistency.
Later @CoderForTheBetter intends to add helper code which could then make this correct or more intuitive, however that's not essential to do immediately because it will not change the actual rotations applied.

**Comment by pgimeno** [comment]
> simply applies Irrlicht 3D rotations as they work in Irrlicht and in MT

How about renaming it to `set_rotation_raw()` and make it accept degrees, then, and make a new `set_rotation()` (or `set_roll/pitch/yaw`) with the correct behaviour and in radians?

**Comment by paramat** [comment]
I don't think there is a bug, as the implementation was discussed in depth and is intentionally done to be consistent with attachments which are also like this. Further code is intended later for more intuitive usage.
However this needs discussion with @CoderForTheBetter

**Comment by CoderForTheBetter** [comment]
Yeah, I had the same idea about using 'raw' somewhere in the name. However, the convention was going to be something like 'set_rotation' and then 'rotate' for the other one. Using 'set_rotation_raw' seems redundant in the case I was just talking about, and it may hint that there could be a normal 'set_rotation' method when there really isn't. Just my opinion on the naming convention.

As for degrees and radians. I don't think anyone wants the API to get cluttered with different methods for different units. In any case, I will be trying to stick with radians since that is what had been agreed upon.

**Comment by paramat** [comment]
@CoderForTheBetter but what about this issue? Do you agree it's a non-issue and can be closed?

**Comment by CoderForTheBetter** [comment]
I don't think there is an issue. Like you said, that PR was merged under the understanding that the rotations would act just like the attachment rotations. No matter what the consequences of that are, this was needed and if people are clever enough they will be able to usefully utilize the new method for their mods. When I have the time (my finals are over the end of next week), I will start on trying to get a PR going that will help make rotations easier to work with.

Sorry if I missed another point in the issue, it's a lot to take in and think about.

**Comment by paramat** [comment]
No problem. I don't think we have an issue open for rotation helper code, and this issue has a useful discussion, so i'll leave it open.

**Comment by CoderForTheBetter** [comment]
I'll use this as a discussion area for now. I found a good link for rotations that I think that we want from the Irrlicht forums: http://irrlicht.sourceforge.net/forum/viewtopic.php?f=1&t=51546

I'm pretty sure we want local axis rotation. It behaves as I would want it to.

I made my own Irrlicht scene and airplane model to test this (pay attention the buttons pressed in the top left and how they might relate to airplane controls): [Video](https://streamable.com/pxwi8)

Using the code provided in that forum topic, it would be very easy to implement this without any edits to the network's protocol. 

The new method would be 'rotate' and it would have its rotations based on local/model space axes. It would take a vector3df that would define the rotation to be applied. Using the other 'set_rotation' method already implemented, it would be possible to reset the rotation to zero if needed, for example.

**Comment by pgimeno** [comment]
Most of the time, they would only use one nonzero axis. Also, the XYZ order is unnatural in Minetest, as discussed. I suggest to have separate `rotate_x()`, `rotate_y()` and `rotate_z()` instead, and let the user decide the order of rotations in case they want to apply several. For example, to set roll, pitch and yaw of an object with a forward vector of Z and an up vector of Y, which I believe is the boats mod's convention, they could do this:

```lua
  object:set_rotation{x=0, y=0, z=0}
  object:rotate_y(yaw)
  object:rotate_x(pitch)
  object:rotate_z(roll)
```

This is going to have several problems, but these are not solvable without modifying the way that Irrlicht's `ISceneNode` handles rotations. In addition to the one I already mentioned near the gimbal lock angle, there will be an accumulated precision problem, where rotating an object over one axis will slowly drift the rotation in the other axes, which will be a problem in e.g. a fan, unless the user protects against that by starting with a fresh rotation every frame.

**Comment by CoderForTheBetter** [comment]
@ClobberXD the controls aren't of any importance as it is just for demonstration, but I do agree with you. I had already recorded it and didn't want to do it again just to change the controls.

@pgimeno I was thinking about the separate functions and I might do just that, but there are some cases where someone might have a forward velocity and a sideways velocity and they will create a vector that is in between the two, for example. To get the resulting rotation, they would have to break the vector into components and apply them separately, and I'm not sure how well that would work.

As for gimble lock and precision, you're right but as you said there is no nice way to deal with this. I hope, that if we add enough helped code for rotations, then people will be able to solve their problems by using their own ideas and the methods provided.

Sorry, if some of the stuff I say does not make sense in this context, I'm relatively new to the ideas of rotations and am actually reading a book about game engines and rotations to get a better understanding. Thanks for your in-depth discussion on rotations, pgimeno.

**Comment by pgimeno** [comment]
@CoderForTheBetter 

tl;dr: On second thought, having a single function perform all three local relative rotations is better, if they are performed in YXZ order and the rotation over each axis is not performed if the angle for that axis is zero. Signs may need adjustment. I'd suggest adding a new function `get_euler()`. There's a separate but related issue mentioned at the bottom.

Full discussion:

> I was thinking about the separate functions and I might do just that, but there are some cases where someone might have a forward velocity and a sideways velocity and they will create a vector that is in between the two, for example. To get the resulting rotation, they would have to break the vector into components and apply them separately, and I'm not sure how well that would work.

Local relative rotation in three axes in XYZ order as suggested would make that goal extremely user-unfriendly. The easiest way to point the nose (face) of an object to the desired direction is to create a function that accepts a forward vector and a roll angle (or a forward vector and an up vector, which is basically matrix form; more advanced, but it also works for objects pointing exactly up). Absent that, you can use yaw, pitch and roll, **IF** they work the way they are designed to work. With intrinsic (=local) XYZ order, they don't, at least in objects that use Z=forwad and Y=up. That order would make X work as a horizontal "yaw" and Y work as a weirdly oriented "pitch", which are as nearly useless as the behaviour described in the OP of this thread. Only Z would work correctly as roll.

I don't know if having Z=forward, Y=up is the standard convention for Minetest objects. It's the one that the Boats mod uses. If it is the gold standard, then the natural order is intrinsic YXZ. With that, you can make an object point wherever you want, by first resetting the rotation, then rotating over the local Y by the desired yaw angle, then over the local X by the desired pitch angle and finally over the local Z by the desired roll angle.

So, the function you suggested that performs a local relative rotation over the three axes, would need to perform Y first, then X and then Z. The drawback will be that it will always spend three matrix rotations even when the user only wants one. The three functions method wouldn't waste those. On the other hand, the three functions method needs to perform expensive and precision-losing transformations from Euler to matrix and back to Euler for every rotation, therefore if the user wants more than one, not only will it waste more calculations, but also lose more precision. So on second thought, it's probably best to have all in one as you suggested, but in YXZ order. Adding a condition to not apply the local rotation for an axis if the angle is zero will get rid of the extra calculations when not needed.

I haven't mentioned signs yet. The signs may need adjustment. Left-handed coordinate systems confuse me, and I can't find the sign convention of Minetest angles in the docs anyway.

Note that despite the name, `get_yaw()` does not return the yaw angle. It returns the Y component of `get_rotation()`, which is a strange beast as described in the OP. Perhaps a function like `get_euler()` could be added, that gets actual pitch, yaw and roll using the boats axis convention, i.e. the one described by rubenwardy. I have already finished the Lua code that converts from `get_rotation()` to pitch/yaw/roll (the function XYZ2ZXY that I left unimplemented in the test code above). Here it is:

```lua
local function XYZ2ZXY(rot)
  local a1, a2, a3 = -rot.x, -rot.y, -rot.z
  local sin, cos, atan2, sqrt = math.sin, math.cos, math.atan2, math.sqrt
  local c1, s1 = cos(a1), sin(a1)
  local c2, s2 = cos(a2), sin(a2)
  local c3, s3 = cos(a3), sin(a3)

  -- matrix from XYZ, Irrlicht style

  local m00, m01, m02 = c2*c3, c2*s3, -s2
  local m10, m11, m12 = s1*s2*c3 - c1*s3, s1*s2*s3 + c1*c3, s1*c2
  local m20, m21, m22 = c1*s2*c3 + s1*s3, c1*s2*s3 - s1*c3, c1*c2

  -- matrix to ZXY
  -- extract Tait-Bryan angles
  a1 = atan2(m01, m11)
  c2 = sqrt(m22*m22 + m20*m20)
  a2 = atan2(-m21, c2)

  -- Shoemake's original
--  a3 = atan2(m20, m22)

  -- Mike Day's replacement
  -- https://www.semanticscholar.org/paper/Extracting-Euler-Angles-from-a-Rotation-Matrix-Day/668137fa4b875d890f446e689eea1e334bcf6bf6
  c1, s1 = cos(a1), sin(a1)
  a3 = atan2(s1*m12 - c1*m02, c1*m00 - s1*m10)

  -- return them in the correct order
  return {x = a2, y = a3, z = a1}
end
```

I'm not sure about the correctness of the output sign convention in this code.

Separate issue:

The docs should probably be adjusted, to mention that X and Y of `set_rotation()` only work as pitch and yaw as long as Z is zero, and that `get_yaw()` doesn't return the yaw if the Z angle is not zero. I'm wondering if it would be best to completely drop the idea of adding any support functions for general rotation, and only support rotations with Z=0. Only advanced users would be able to use them properly otherwise, due to the cumulative precision loss, which defeats the purpose of adding functions to simplify their use. The mess created by Irrlicht's node rotation method is big.

**Comment by paramat** [comment]
> The docs should probably be adjusted, to mention that X and Y of set_rotation() only work as pitch and yaw as long as Z is zero

I agree, i had to make this clear with my boats code to avoid user confusion.

**Comment by pgimeno** [comment]
After more thought, I think the best solution to get rid of the mess would be to override the virtual method `getRelativeTransformation` in `ISceneNode` to make x work as pitch, y as yaw and z as roll as desired. Besides solving most major problems, that would then open the gate to implementing a better fix in future, like storing the data in matrix form internally. This needs to be done in a major version, and 5.0.0 looks like the best one since `set_rotation()` and `get_rotation()` have just been introduced.

I can propose a patch, but I no longer send PRs to MS-GitHub so that will be a problem for now as is the case with other patches I've sent. If I submit it to GitLab, would developers be able to review it there?

With that patch:

* `set_rotation()` would set, and `get_rotation()` would get, real pitch, yaw and roll.
* There would be no precision loss in the other angles for incremental rotation in one of the axes.
* The angles would be intuitive already.
* The interpolation would have its special case only when the object is looking straight up, rather than horizontally. That's less frequent (hence less likely to cause trouble) and easier to understand and assimilate as an engine limitation.

A function to perform local relative rotation in the three axes in YXZ order would still be useful.

**Comment by pgimeno** [comment]
Disregard that last post, it's not viable to fix Irrlicht without overriding a big bunch of Irrlicht classes and making some major changes to the code. When (or rather, if) Minetest decides to fork Irrlicht to keep a local patched version, the rotations will have a chance of being fixed.

Meanwhile, dropping support for rotations with Z angle != 0 seems the best option. Either tell users "do it at your own risk, it doesn't work sanely and it will break in future" or plainly force it to 0 in `set_rotation()`.

**Comment by paramat** [comment]
I would prefer strongly recommending Z = 0 for intuitive YX yaw-pitch behaviour, but still allowing Z rotation for those who can work with it, for one-time rotation setting, or for use with the future helper functions.

**Comment by CoderForTheBetter** [comment]
I agree with paramat, I don't see any reason for not allowing Z rotation. However, the docs should be added to, so that people know what is going on.

When I have more time I will come back to this discussion and put more thought into, especially some of the ideas you mention in this [comment](https://github.com/minetest/minetest/issues/7927#issuecomment-445346938)

**Comment by paramat** [comment]
No problem and no rush.

**Comment by pgimeno** [comment]
There are several reasons why I suggested to drop support:

- As soon as Z != 0, `get_yaw()` and `set_yaw()` will break, and not report/set the actual yaw (they get/set the Y angle of the broken Eulers described in the OP). Any fix for this without changing Irrlicht is going to be subject to precision issues (*).
- If Z != 0 is allowed, and a proper fix is written in future, supporting this broken method for backward compatibility is going to be a pain, if possible at all (due to precision issues again).
- Allowing Z != 0 means that a proper fix will require another protocol break.

(*) I've measured the precision loss with the ZXY2XYZ and XYZ2XZY functions I was using. Starting with a boat at zero rotation, i.e. pointing North, and then continuously reading the rotation, changing the yaw and setting the rotation again, the Z angle drifted by 1 degree every 1400 turns approximately.

**Comment by paramat** [comment]
Ok thanks, i don't fully understand the issues, still getting my head around it.

The important thing is, what else needs to be done, if anything, before 5.0.0, considering the network compatibility breakage? Or are we ok just having the raw unintuitive Irrlicht rotation we have now? We could document 3D rotation as unstable for 5.0.0 but leave it usable while we continue to consider what to do.

**Comment by pgimeno** [comment]
My suggestion: to change `set_rotation()` and `get_rotation()` to `set_pitch()` and `get_pitch()`, (since there's already `set_yaw()` and `get_yaw()`), to keep `set_rotation()`/`get_rotation()` available for the future. Also to not implement `rotate`, as it allows changing Z and thus making `get_yaw()` misbehave.

Alternatively, leave `get/set_rotation()` in, and implement `rotate()`, keeping in mind that `get/set_yaw()` will break, and that in a future (6.0.0) version, they may change behaviour in an incompatible way.

I'm going to report this in the Irrlicht forums and see what they say. With some luck, this may get fixed in a newer Irrlicht version that Minetest can require. Of course, that process can take years until a fixed Irrlicht is available in the major distributions.

**Comment by CoderForTheBetter** [comment]
If you go way back to an issue I made before the set_rotation PR. I and stujones talked about doing exactly that briefly.

**Comment by pgimeno** [comment]
I've taken a look. It seems the issue with roll has been already discussed, and I haven't been the only one suggesting to drop support for roll (Z!=0) as a possibility.

The suggestion by @stujones11 seems more wasteful to me, in that it consisted of adding three `EmptySceneNode`s one for each angle. The thread in the Irrlicht forums I linked discusses basically the same problems as in this thread, namely the issues of the choice of axis convention in relation to Eulers. The solution in the linked post is to add a special node that stores a whole 4x4 transformation matrix, instead of a normal scene node that stores the rotation as XYZ Eulers. That gives maximum flexibility with minimum precision loss, as the matrix can be manipulated directly through the built-in functions without needing to convert from/to Euler in the middle.

I already have WIP code using that idea: https://gitlab.com/pgimeno/minetest/commits/pg-fix-roll

It can be imported into an existing Minetest repo with:

```
git remote add pgimeno https://notabug.org/pgimeno/minetest
git fetch pgimeno pg-fix-roll
git checkout pg-fix-roll
```

It's working, and makes roll/pitch/yaw work as designed. The matrix stores position and rotation, but not scale, which is stored in the child node. There's a todo list in the commit message.

**Comment by CoderForTheBetter** [comment]
I have some time to address and read your comments here.

> As soon as Z != 0, get_yaw() and set_yaw() will break, and not report/set the actual yaw (they get/set the Y angle of the broken Eulers described in the OP). Any fix for this without changing Irrlicht is going to be subject to precision issues (*).

I can't believe I didn't mention this anywhere and that I did not address the issue clearly (if at all). My thought going into the original PR was that support for 'set_yaw' and 'get_yaw' would be dropped, but then every mod in existence that uses those would have to be updated, not ideal, so I kept it without ever addressing the issue of it breaking. Thank you for being so detailed in your discussion.

> Alternatively, leave get/set_rotation() in, and implement rotate(), keeping in mind that get/set_yaw() will break, and that in a future (6.0.0) version, they may change behaviour in an incompatible way.

That is exactly what I have been going for. I would much prefer this route and I hope others with share their opinions. Of course, we would need to update the docs to address these issues.

As for the branch you are working on, I remember looking at the same Irrlicht forum thread a while ago and then deciding to first get the network code working, and then move on to 'rotate' without first thinking of the consequnces. I wouldn't have thought of the errors you have addressed though (precision and different Z).

If the extra node is not too _hackish_ (just seems so to me, maybe not to others) then I hope we can take that route as it seems to be the most useful and reliable.

I look forward to testing your branch when I get more time!

**Comment by pgimeno** [comment]
In my view, the hack is the "official workaround" of their known limitations. To quote from the forum post:

>> I think it is a good idea in the future to allowing directly set the node to quaternion or transformation matrice, since there is already a get matrix function.
>
> It's an idea that has been floating around a few times already. But don't expect it to be a quick hack when you want to do it right.
> IDummyTransformationSceneNode did go in this direction - allowing to add nodes which work directly with matrices.

**Comment by paramat** [comment]
Is there anything that needs doing here before release of 5.0.0? As in, anything that would cause a network compatibility breakage? Wondering if this is a blocker.

**Comment by pgimeno** [comment]
Yes, the network protocol is currently transmitting the broken Eulers, and if it stays that way, it can't be fixed until 6.0.0. The patch I linked to is still WIP. In its current state, that patch transmits correct Eulers, but the best fix would be to transmit the rotation as a matrix or quaternion instead.

I still think that the behaviour of `set_yaw()` and `get_yaw()` is broken when Z!=0, as is the behaviour of `set_rotation()` and `get_rotation()` for nonzero Z as well, and that therefore this issue is a bug and not an enhancement.

This can be postponed to the 5.0.1+ development cycle if:

- `set_rotation` and `get_rotation` are removed as functions; possibly, `set_pitch_yaw()` and `get_pitch_yaw()` are added in their place, that only change X and Y but leave Z as zero.
- The network protocol is assumed to use Eulers (which has problems on its own, but hopefully minor), and conversion code is added to convert from broken to correct Eulers on send and vice versa on receive.

The advantage of such an approach is that the patch to do this is much simpler, and true `set_rotation()` and `get_rotation()` can be re-added post-5.0.0 without breaking the protocol, as soon as the more complete patch (which will be harder to review) is included. In the transition period (5.0.0) there will be a precision loss in the conversion, but not an *accumulated* precision loss, which is the fatal one. More importantly, there can also be gimbal lock issues at 90° yaw and zero or near zero pitch.

I have the conversion itself already written (I've posted most of it in this thread) but not the patch. I can prepare it if that's deemed the best approach.

**Comment by pgimeno** [comment]
Alas, no, the above solution isn't possible. I've tried, but the conversion function returns an angle for the actual pitch (the one that is used as yaw by Minetest) in the range [-90°, 90°], so it only works right in a semicircle. I didn't think of that.

I'll keep working on the bigger patch.

**Comment by paramat** [comment]
Will your work break attachment rotations which have been 3D rotations for a long time?

**Comment by pgimeno** [comment]
No, but I will need some way to test them. Any idea how?

**Comment by pgimeno** [comment]
The patch is complete. Done:

- [x] Test multiplayer.
- [x] Test wielded items.
- [x] Test armour.
- [x] Test attaching an object to another.
- [x] Test support for attachment rotations with Z!=0.
- [x] Test animated meshes.
- [x] Move setPitchYawRollRad() and setPitchYawRoll() to util/numeric.cpp|h
- [x] Check camera.cpp/h to see if there's anything to change there.
- [x] Check whether the interpolator is adequate for the new ZXY order.
- [x] Add internal function to convert the matrix back to Eulers.
- [x] Fix sprites (they were broken in earlier versions of the patch).
- [x] Explain the new behaviour in the docs for `set_rotation`.

Not done:

- Considered sending rotation as matrix through the network. Decided against, due to the cost in data transmission (if sent as matrix) or in postprocessing (if sent as quaternion). Also, the algorithm being used for Euler conversion has a guard against a particularly evil gimbal lock effect, near the pole (at the cost of a bit of precision), so it should be safe.
- Bump protocol version. Apparently it's not that necessary for unstable releases. Added a note about the yaw -> rotation change to networkprotocol.h instead, as suggested by Krock on IRC.

Notes on the test procedure:
- Attach player to object was tested with MTG's boats. Rotations were tested with Paramat's change to MTG's boats taken from https://github.com/minetest/minetest/pull/7395#issuecomment-426861772.
- Attach object to player, as well as attachment rotations, were tested with a modified version of [wield3d](https://github.com/stujones11/wield3d) (modified the angles in location.lua to test with all angles set to nontrivial rotations).
- Attach object to object was tested with [boost_cart](https://forum.minetest.net/viewtopic.php?t=10172), by dropping an item into a cart. I used a modified version of the original pre-MTG [carts](https://forum.minetest.net/viewtopic.php?t=2451) mod (modified to make it work acceptably with 5.0), not sure if that makes a difference.
- Armour was tested with [3d_armor](https://forum.minetest.net/viewtopic.php?t=4654).
- Animated meshes were tested with [nssm](https://forum.minetest.net/viewtopic.php?t=11813) + [mobs_redo](https://forum.minetest.net/viewtopic.php?t=9917).

**Comment by pgimeno** [comment]
One additional problem I've just found in #7395, which the above patch doesn't fix, is that the saved rotations for entities are broken for existing worlds. I guess that's a consequence of the change in the serialization format. The old yaw goes where the new pitch is, so the boats migrated from stable-0.4 get this appearance:

![boat-migrated](https://user-images.githubusercontent.com/4267396/50035140-c06ed600-0000-11e9-87a2-1499588dc8e9.png)

I don't know how to handle this.

**Comment by paramat** [comment]
Hm that is bad, a blocker alone.
Could yaw/y-rotation continue to be used over the newwork and we add 2 separate floats for x and z rotation?

**Comment by pgimeno** [comment]
Fixed the last bit. Set as a merge request to my own repository, to hopefully allow review: https://gitlab.com/pgimeno/minetest/merge_requests/3

**Comment by pgimeno** [comment]
The XYZ Eulers issue has been reported in Irrlicht here: http://irrlicht.sourceforge.net/forum//viewtopic.php?f=7&t=52414

I wouldn't expect a new version of Irrlicht with a fix for that issue to be released before 5.0.0. Furthermore, if it's fixed on Irrlicht, the `IDummyTransformationSceneNode` hack can be reverted at any time without a protocol break.

**Comment by paramat** [comment]
Shall i make a PR from your patch for GitHub?
Do i use the 'plain diff' and 'git am name.patch' it to a PR branch?

**Comment by pgimeno** [comment]
You can fetch the branch into your regular MT repo: https://github.com/minetest/minetest/issues/7927#issuecomment-446046092

Then you can submit the branch as a PR, in the same way as if you authored it (you would use the remote 'pgimeno' to fetch the branch, and I imagine that you use the remote 'origin' to push it). No need to fiddle with git am.

**Comment by paramat** [comment]
Ok thanks but it's best i leave this to someone else as i'm not good with git.

**Comment by paramat** [comment]
The lack of a PR is, i suspect, causing your patch to be less visible, we need someone to open a PR with your patch.

**Comment by paramat** [comment]
> Try to imagine the boat as an airplane with the nose pointing to the starboard of the boat, the wings crossing the table perpendicularly, and the bottom of the plane where the boat engine is. Then try to imagine that the water is a vertical wall, and it will all make sense.
With up/down (shift/space) you control the yaw over the Z axis, with left/right (A/D) the pitch, and with fwd/back (W/S) the roll.

Tried this now in MT master and can confirm rotations are now intuitive and have the frame-of-reference order ZYX: yaw-pitch-roll.

**Comment by paramat** [comment]
Re-reading the thread i understand the issues better now.
